File: main/src/main/java/cgeo/geocaching/command/MoveToListAndRemoveFromOthersCommand.java
package cgeo.geocaching.command;

import cgeo.geocaching.models.Geocache;
import cgeo.geocaching.storage.DataStore;
import cgeo.geocaching.utils.IgnoreListUtils;

import android.app.Activity;

import androidx.annotation.NonNull;

import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public abstract class MoveToListAndRemoveFromOthersCommand extends MoveToListCommand {

    private final Map<String, Set<Integer>> oldLists = new HashMap<>();

    protected MoveToListAndRemoveFromOthersCommand(@NonNull final Activity context, @NonNull final Geocache cache) {
        this(context, Collections.singleton(cache));
    }

    protected MoveToListAndRemoveFromOthersCommand(@NonNull final Activity context, @NonNull final Collection<Geocache> caches) {
        super(context, caches, -1);
    }

    @Override
    protected void doCommand() {
        for (final Geocache cache : getCaches()) {
            oldLists.put(cache.getGeocode(), new HashSet<>(cache.getLists()));
        }
        final Set<String> wasOnIgnoreList = IgnoreListUtils.snapshotIgnoreListMembership(getCaches());
        DataStore.saveLists(getCaches(), Collections.singleton(getNewListId()));
        IgnoreListUtils.reflectMembershipChange(getCaches(), wasOnIgnoreList);
    }

    @Override
    protected void undoCommand() {
        final Set<String> wasOnIgnoreList = IgnoreListUtils.snapshotIgnoreListMembership(getCaches());
        for (final Geocache cache : getCaches()) {
            final Set<Integer> listIds = oldLists.get(cache.getGeocode());
            if (listIds != null) {
                DataStore.saveLists(getCaches(), listIds);
            }
        }
        IgnoreListUtils.reflectMembershipChange(getCaches(), wasOnIgnoreList);
    }

}

File: main-kotlin/src/main/kotlin/cgeo/geocaching/command/MoveToListAndRemoveFromOthersCommand.kt
// Auto-converted from Java to Kotlin
// WARNING: This code requires manual review and likely has compilation errors
// Please review and fix:
// - Method signatures (parameter types, return types)
// - Field declarations without initialization
// - Static members (use companion object)
// - Try-catch-finally blocks
// - Generics syntax
// - Constructors
// - And more...

package cgeo.geocaching.command

import cgeo.geocaching.models.Geocache
import cgeo.geocaching.storage.DataStore

import android.app.Activity

import androidx.annotation.NonNull

import java.util.Collection
import java.util.Collections
import java.util.HashMap
import java.util.HashSet
import java.util.Map
import java.util.Set

abstract class MoveToListAndRemoveFromOthersCommand : MoveToListCommand() {

    private final Map<String, Set<Integer>> oldLists = HashMap<>()

    protected MoveToListAndRemoveFromOthersCommand(final Activity context, final Geocache cache) {
        this(context, Collections.singleton(cache))
    }

    protected MoveToListAndRemoveFromOthersCommand(final Activity context, final Collection<Geocache> caches) {
        super(context, caches, -1)
    }

    override     protected Unit doCommand() {
        for (final Geocache cache : getCaches()) {
            oldLists.put(cache.getGeocode(), HashSet<>(cache.getLists()))
        }
        DataStore.saveLists(getCaches(), Collections.singleton(getNewListId()))
    }

    override     protected Unit undoCommand() {
        for (final Geocache cache : getCaches()) {
            val listIds: Set<Integer> = oldLists.get(cache.getGeocode())
            if (listIds != null) {
                DataStore.saveLists(getCaches(), listIds)
            }
        }
    }

}
